dataflow.py

```diff
@@ -1,81 +1,98 @@
+    async with websockets.connect(url) as websocket:
+        msg = json.dumps(
+            {
+                "type": "subscribe",
+                "product_ids": [product_id],
+                "channels": ["level2_batch"],
+            }
         )
+        await websocket.send(msg)
         # The first msg is just a confirmation that we have subscribed.
-        print(self.ws.recv())
+        await websocket.recv()
 
-    def next(self):
-        return self.ws.recv()
+        while True:
+            msg = await websocket.recv()
+            yield (product_id, json.loads(msg))
 
 
-class CoinbaseInput(DynamicInput):
-    PRODUCT_IDS = ["BTC-USD", "ETH-USD", "SOL-USD"]
+class CoinbasePartition(StatefulSourcePartition):
+    def __init__(self, product_id):
+        agen = _ws_agen(product_id)
+        self._batcher = batch_async(agen, timedelta(seconds=0.5), 100)
 
-    def build(self, worker_index, worker_count):
-        prods_per_worker = int(len(self.PRODUCT_IDS) / worker_count)
-        product_ids = self.PRODUCT_IDS[
-            int(worker_index * prods_per_worker) : int(
-                worker_index * prods_per_worker + prods_per_worker
-            )
-        ]
-        return CoinbaseSource(product_ids)
+    def next_batch(self):
+        return next(self._batcher)
 
+    def snapshot(self):
+        return None
 
-flow = Dataflow()
-flow.input("input", CoinbaseInput())
 
-flow.map(json.loads)
-# {'type': 'l2update', 'product_id': 'BTC-USD', 'changes': [['buy', '36905.39', '0.00334873']], 'time': '2022-05-05T17:25:09.072519Z'}
+@dataclass
+class CoinbaseSource(FixedPartitionedSource):
+    product_ids: List[str]
 
+    def list_parts(self):
+        return self.product_ids
 
-def key_on_product(data):
-    return (data["product_id"], data)
+    def build_part(self, step_id, for_key, _resume_state):
+        return CoinbasePartition(for_key)
 
 
-flow.map(key_on_product)
-# ('BTC-USD', {'type': 'l2update', 'product_id': 'BTC-USD', 'changes': [['buy', '36905.39', '0.00334873']], 'time': '2022-05-05T17:25:09.072519Z'})
+@dataclass(frozen=True)
+class OrderBookSummary:
+    bid_price: float
+    bid_size: float
+    ask_price: float
+    ask_size: float
+    spread: float
 
 
-class OrderBook:
-    def __init__(self):
-        self.bids = {}
-        self.asks = {}
+@dataclass
+class OrderBookState:
+    bids: OrderedDict = field(default_factory=OrderedDict)
+    asks: OrderedDict = field(default_factory=OrderedDict)
+    bid_price: Optional[float] = None
+    ask_price: Optional[float] = None
 
     def update(self, data):
-        if self.bids == {}:
-            self.bids = {float(price): float(size) for price, size in data["bids"]}
-            # The bid_price is the highest priced buy limit order.
-            # since the bids are in order, the first item of our newly constructed bids
-            # will have our bid price, so we can track the best bid
+        if len(self.bids) <= 0:
+            self.bids = OrderedDict(
```

Comment:
I think OrderedDict isn't necessary in >3.7 since they are ordered by default.

Reply:
Modified the code - the logic is a bit different and I am nervous the new logic doesn't return the exact results as with the ordered one. 

I see 2 options

1. create a PR in the bytewax/bytewax repo under examples to match both
2. leave this blog with the original implementation
https://github.com/bytewax/crypto-orderbook-app/pull/2/commits/d13c13f5d6eec472ff28f7c9d784463006bfbcf0